src/Buffer.mo

```diff
@@ -1,166 +1,1754 @@
+    ///
+    /// Runtime: O(size * log(size))
+    ///
+    /// Space: O(size)
+    public func sort(compare : (X, X) -> Order.Order) {
+      // Stable merge sort in a bottom-up iterative style
+      if (count == 0) {
+        return;
+      };
+      let scratchSpace = Prim.Array_init<?X>(count, null);
+
+      let countDec = count - 1 : Nat;
+      var currSize = 1; // current size of the subarrays being merged
+      // when the current size == count, the array has been merged into a single sorted array
+      while (currSize < count) {
+        var leftStart = 0; // selects the current left subarray being merged
+        while (leftStart < countDec) {
+          let mid : Nat = if (leftStart + currSize - 1 : Nat < countDec) { leftStart + currSize - 1 } else { countDec };
+          let rightEnd : Nat = if (leftStart + (2 * currSize) - 1 : Nat < countDec) { leftStart + (2 * currSize) - 1 } else { countDec };
+    
+          // Merge subarrays elements[leftStart...mid] and elements[mid+1...rightEnd]
+          var left = leftStart;
+          var right = mid + 1;
+          var nextSorted = leftStart;
+          while (left < mid + 1 and right < rightEnd + 1) {
+            let leftOpt = elements[left];
+            let rightOpt = elements[right];
+            switch (leftOpt, rightOpt) {
+              case(?leftElement, ?rightElement) {
+                switch(compare(leftElement, rightElement)) {
+                  case (#less or #equal) {
+                    scratchSpace[nextSorted] := leftOpt;
+                    left += 1;
+                  };
+                  case (#greater) {
+                    scratchSpace[nextSorted] := rightOpt;
+                    right += 1;
+                  }
+                };
+              };
+              case (_, _) { // only sorting non-null items
+                Prim.trap("Malformed buffer in sort")
+              };
+            };
+            nextSorted += 1;
+          };
+          while (left < mid + 1) {
+            scratchSpace[nextSorted] := elements[left];
+            nextSorted += 1;
+            left += 1;
+          };
+          while (right < rightEnd + 1) {
+            scratchSpace[nextSorted] := elements[right];
+            nextSorted += 1;
+            right += 1;
+          };
+
+          // Copy over merged elements
+          var i = leftStart;
+          while (i < rightEnd + 1) {
+            elements[i] := scratchSpace[i];
+            i += 1;
+          };
+
+          leftStart += 2 * currSize;
+        };
+        currSize *= 2;
+      }
     };
 
-    /// Returns an `Iter` over the elements of this buffer.
+    /// Returns an Iterator (`Iter`) over the elements of this buffer.
+    /// Iterator provides a single method `next()`, which returns
+    /// elements in order, or `null` when out of elements to iterate over.
+    /// 
+    /// Runtime: O(1)
+    ///
+    /// Space: O(1)
     public func vals() : { next : () -> ?X } = object {
```

Comment:
Actually, I'm not sure this implementation is so dangerous, is it? Although it doesn't detect concurrent modification, it always consults the current `count` and backing array `elements` and doesn't iterate over any older version. This is because the iterator objects closes over the locations of count and elements, not the current value.

@ggreif

Reply:
Testing tagging luc @luc-blaeser

Reply:
Right, and I think there won't be out of bound traps (since count < elements.size() is an invariant).

To detect concurrent modification, Luc's suggestion was to keep a version variable that gets incremented (and possibly moduloed) by any method mutating the logical list, and to check against this version in every call to next().

But maybe "safer iterators" can be a later ticket/feature that adds this safety to a bunch of data structures. At minimum it will be a breaking change, where this current PR is backwards compatible.